### ISSUE TITLE
Arithmetic in Axes and leading dimensions in tjax.ArrayN

### ISSUE THREAD

**Issue by aidanscannell** [deepmind/tensor_annotations#2]
Hello, first of all, thanks for a neat little library, I've been loving it. 

I have two features that I do not think are currently implemented and I would definitely use. I am wondering if there's scope to implement them? Maybe I could give them a go if they're something you think are worth adding?  

1. Is it possible to perform arithmetic in the array axes. For example, if we had data with shape `[InputDim]` and the output of the function should return a 1D array with shape 2*InputDim. I would like to use something like this,
```   
def my_func(x: tjax.Array1[InputDim]) -> tjax.Array1[InputDim + InputDim]:
```
Is there a way to implement this?

2. In JAX I use the `jax.vectorize` function to handle auto batching of my functions and the output of these function can have any number of leading dimensions. Is there a way to specify a batched array which can have any number of dimensions equal to or more than some number? For example,
```
tjax.BatchedArray2[InputDim, InputDim] = Union[tjax.Array3[Batch1, InputDim, InputDim], tjax.Array4[Batch1, Batch2, InputDim, InputDim], tjax.Array5[Batch1, Batch2, Batch3, InputDim, InputDim], ...]
```

Cheers,
Aidan

**Comment by mrahtz** [comment]
Thanks for the kind words, and sorry for the crazy slow reply - had a deadline and was in "ignore everything else possible" mode 😅

> Is it possible to perform arithmetic in the array axes. For example, if we had data with shape [InputDim] and the output of the function should return a 1D array with shape 2*InputDim.

Unfortunately, this isn't possible for the time being, but _may_ be possible in the future.

The reason it's not possible right now is that there's no concept of arithmetic in the Python type system, so type checkers wouldn't understand what such expressions meant. However, there _is_ some work on making this possible - e.g. see [this presentation](https://docs.google.com/presentation/d/16VGNfVo6Kxk9lBhm1cx1-Jr19dqESyy10WkhtMgUeDY/edit#slide=id.g9706e4c7e7_0_146) and [this talk](https://youtu.be/ld9rwCvGdhc?t=11479). There _is_ some preliminary support for this in Pyre already, but it's otherwise still fairly early work, so it won't be supported by type checkers in general any time soon.

There's also a second issue, which is a consequence of a design decision - in `tensor_annotations` we deliberately decided that labels like `InputDim` should represent semantic types rather than actual values. That is, in the `tensor_annotations` way of looking at things, `InputDim * 2` would look like `int * 2`, which wouldn't make any sense - what would it mean to multiply a type by 2? We decided things this way for reasons that are currently best written up at https://www.python.org/dev/peps/pep-0646/#appendix-a-shape-typing-use-cases but we may change this in the future depending on which use case people find more valuable.

> In JAX I use the jax.vectorize function to handle auto batching of my functions and the output of these function can have any number of leading dimensions. Is there a way to specify a batched array which can have any number of dimensions equal to or more than some number?

This is also another thing we don't support very well yet, and is on our priority list of things to figure out how to do better. Probably the best way you do it as you suggest, but with a type variable in place of `InputDim`, and making the union go up to the maximum number of batch dimensions you expect. The following is a bit hacky but seems to do the right thing in Mypy:

```python
from functools import partial
from typing import Any, NewType, TYPE_CHECKING, TypeVar, Union

import jax
import jax.numpy as jnp
from tensor_annotations import axes
from tensor_annotations.axes import Axis, Batch
from tensor_annotations.jax import Array2, Array3, Array4
import tensor_annotations.jax as tjax

Batch1 = NewType('Batch1', Axis)
Batch2 = NewType('Batch2', Axis)
InputDim = NewType('InputDim', Axis)

T = TypeVar('T', bound=Axis)
BatchedArray1 = Union[
    Array2[Any, T],
    Array3[Any, Any, T],
]
BatchedArray2 = Union[
    Array3[Any, T, T],
    Array4[Any, Any, T, T],
]

@partial(jnp.vectorize, signature='(k)->(k,k)')
def foo(x):
    return x[..., None] + x[None, ...]

def vectorized_foo(x: BatchedArray1[T]) -> BatchedArray2[T]:
    return foo(x)

def bar(x: BatchedArray2):
    pass

x1: Array2[Batch1, InputDim] = jnp.zeros((3, 5))
x2: Array3[Batch1, Batch2, InputDim] = jnp.zeros((3, 5, 7))

y1 = vectorized_foo(x1)
y2 = vectorized_foo(x2)

if TYPE_CHECKING:
    reveal_type(x1)
    reveal_type(x2)
    reveal_type(y1)
    reveal_type(y2)

bar(y1)
bar(y2)
```

The hacky thing here is that we don't currently have a stub for `jnp.vectorize` so we have to force the signature for the vectorized `foo` using `vectorized_foo`. I've added a TODO for trying to add stubs for vectorize directly.

(Also, sorry that we only go up to Array4 currently. Should be able to fix this soon.)

**Comment by aidanscannell** [comment]
Thanks for the great response and all of the links, I will be checking them all out! Using BatchedArrayN and the `vectorized_foo` seems like a good solution for now so I will go ahead and use that. I look forward to the `jnp.vectorize` stub 😄

**Comment by mrahtz** [comment]
Sure! I'll leave this open for now to remind me to try implementing `jnp.vectorize` stubs.
